packages/yarnpkg-pnpify/sources/hoist.ts

```diff
@@ -283,40 +344,32 @@ const cloneTree = (tree: HoisterTree): HoisterWorkTree => {
  * @param tree stripped down hoisted package tree clone
  */
 const shrinkTree = (tree: HoisterWorkTree): HoisterResult => {
-  const treeCopy = new Map();
-  const idMap = new Map();
-
-  const addNode = (srcNodeId: NodeId, parentNode?: HoisterResultNode): string => {
-    const node = tree.get(srcNodeId)!;
-    const {name, reference, locator, deps, depLocators, depNameToId, hoistedDepNames, peerNames, ...meta} = node;
-    let locatorMap = idMap.get(locator);
-    if (!locatorMap) {
-      locatorMap = new Map();
-      idMap.set(locator, locatorMap);
-    }
-    let dstNodeId = srcNodeId === '.' ? '.' : locatorMap.get(srcNodeId);
-    if (!dstNodeId) {
-      dstNodeId = `${reference === '' ? name : locator}${locatorMap.size === 0 ? '' : `$${locatorMap.size}`}`;
-      locatorMap.set(srcNodeId, dstNodeId);
-    }
-    const newNode = {
-      name, reference, deps: new Set<NodeId>(), ...meta,
+  const treeCopy: HoisterResult = {
+    name: tree.name,
+    references: new Set(tree.references),
+    deps: new Set(),
+  };
+
+  const addNode = (node: HoisterWorkTree, parentNode: HoisterResult, parents: Set<HoisterWorkTree>) => {
+    if (parents.has(node))
+      return;
+
+    const {name, references} = node;
+    const resultNode = {
+      name, references, deps: new Set<HoisterResult>(),
     };
-    if (parentNode)
-      parentNode.deps.add(dstNodeId);
-
-    treeCopy.set(dstNodeId, newNode);
-    for (const depId of node.deps) {
-      const depNode = tree.get(depId)!;
-      if (!peerNames.has(depNode.name)) {
-        addNode(depId, newNode);
+
+    parentNode.deps.add(resultNode);
+
+    for (const dep of node.deps.values()) {
+      if (!node.peerNames.has(dep.name)) {
+        addNode(dep, resultNode, new Set(parents).add(node));
```

Comment:
if speed becomes a concern again, maybe doing a `parents.add(node)` prologue and `parents.delete(node)` epilogue would do the trick (instead of copying `parents` multiple times)?

Reply:
Speed is certainly not affected by these micro optimizations. Speed is affected mainly by the heuristic - how we deduplicate dependency graph. If heuristic is good we will have good speed, if it is not - the time might exponentially grow with the growth of the graph (which is the case with current implementation in master)

Reply:
This parents checking is just a quick way to move forward. I know it is not efficient, not because of garbage collection, however, but because this algorithm has huge number of node visits. Now I have time to write it more efficiently, which is by utilizing circular references, which should be much faster.